anndata/tests/helpers.py

```diff
@@ -337,6 +418,14 @@ def are_equal_dataframe(a, b, exact=False, elem_name=None):
     )
 
 
+@assert_equal.register(AwkArray)
+def assert_equal_awkarray(a, b, exact=False, elem_name=None):
+    from anndata.compat import awkward as ak
+
+    assert a.type == b.type, "type mismatch"
+    assert ak.to_json(a) == ak.to_json(b)
```

Comment:
I don't think the json values are good to compare, since I expect floating point values will be truncated, and I don't know what to expect from key ordering. Maybe `ak.to_list` instead?

Also, these should use `format_msg(elem_name)` in their assertion errors so we can see the key for the element name.

Reply:
The key ordering is guaranteed and floating point values have a lot of digits (10ish?), but you're right that it would make more sense to compare at the level of `to_list`. That's what we do in our own tests: we check types and `to_list` output, which is not to say that the arrays being compared have identical data structures, but it's the level of consistency that we want. It gives us the flexibility to change memory layouts without having all of the tests break.

Reply:
Are there any other equality testing function used for the tests in awkward array?

I think I would also like to check whether types were equivalent, like that strings were string not categorical. Maybe this would be through `.layout.__eq__`?

Reply:
Categorical strings have a different type than non-categorical strings. We're just following Datashape, in which [categorical](https://datashape.readthedocs.io/en/latest/types.html#categorical) is represented by a different type string, so we're calling it distinct. (At least, they should; I'll ask: https://github.com/scikit-hep/awkward/pull/1671#issuecomment-1237477679.)

For detailed metadata comparison, there's `a.layout.form == b.layout.form`.

For reference, we haven't yet decided what `ak.array_equal` ought to do: https://github.com/scikit-hep/awkward/issues/1105 . There are many different kinds of equality one might want with these.

That's why I pointed out that the kind of thing we typically care about in our tests is `type` and `to_list` equality. (That's what `ak.array_equal` should probably do by default, at least.)

Reply:
updated to `to_list()` and we were already comparing types anyway :heavy_check_mark: